src/implicit.rs

```diff
@@ -27,118 +26,200 @@ pub fn resolve_weak(
     sequence: &IsolatingRunSequence,
     processing_classes: &mut [BidiClass],
 ) {
-    // FIXME (#8): This function applies steps W1-W6 in a single pass.  This can produce
-    // incorrect results in cases where a "later" rule changes the value of `prev_class` seen
-    // by an "earlier" rule.  We should either split this into separate passes, or preserve
-    // extra state so each rule can see the correct previous class.
-
-    // FIXME: Also, this could be the cause of increased failure for using longer-UTF-8 chars in
-    // conformance tests, like BidiTest:69635 (AL ET EN)
-
-    let mut prev_class = sequence.sos;
+    // Note: The spec treats these steps as individual passes that are applied one after the other
+    // on the entire IsolatingRunSequence at once. We instead collapse it into a single iteration,
+    // which is straightforward for rules that are based on the state of the current character, but not
+    // for rules that care about surrounding characters. To deal with them, we retain additional state
+    // about previous character classes that may have since been changed by later rules.
+
+    // The previous class for the purposes of rule W4/W6, not tracking changes made after or during W4.
+    let mut prev_w46_class = sequence.sos;
+    // The previous class for the purposes of rule W5.
+    let mut prev_w5_class = sequence.sos;
```

Comment:
But also for the purposes of one chunk of rule W6.

Instead of naming them as the class _for_ a rule, I would prefer if they were named as the class before (or, if you prefer, after) a rule, thus `previous_class_before_w5` or `previous_class_pre_w5` or `previous_class_after_w4` or `previous_class_post_w4`, likewise for the others.
(Also, in this context `prev` confused me somewhat because I was thinking it might mean `pre`, which is how I was thinking of these classes; that abbreviation may be OK if we have a `before`/`pre`/`after`/`post` in the identifier as well.)